File: app/extractor.py
from __future__ import annotations

import html
import ipaddress
import os
import re
import socket
from dataclasses import asdict, dataclass
from typing import Iterable
from urllib.parse import urljoin, urlparse

import httpx
from bs4 import BeautifulSoup


VIDEO_EXTENSIONS = ("mp4", "m3u8", "webm", "mov", "m4v", "ogv")
REQUEST_TIMEOUT = 12.0
MAX_HTML_BYTES = 5 * 1024 * 1024

ABSOLUTE_VIDEO_RE = re.compile(
    rf"(?P<url>(?:https?:)?//[^\s\"'<>]+?\.(?:{'|'.join(VIDEO_EXTENSIONS)})(?:\?[^\s\"'<>]*)?)",
    re.IGNORECASE,
)
RELATIVE_VIDEO_RE = re.compile(
    rf"(?P<url>(?:/|\.\.?/)[^\s\"'<>]+?\.(?:{'|'.join(VIDEO_EXTENSIONS)})(?:\?[^\s\"'<>]*)?)",
    re.IGNORECASE,
)

MEDIA_ATTRS = (
    "src",
    "href",
    "data-src",
    "data-url",
    "data-video",
    "data-video-url",
    "data-hls",
    "data-file",
    "content",
)
EMBED_TAGS = {"iframe", "embed", "object"}
DIRECT_MEDIA_TAGS = {"video", "source"}


@dataclass(frozen=True)
class VideoCandidate:
    url: str
    kind: str
    source: str


class ExtractionError(ValueError):
    pass


def validate_fetch_url(url: str) -> str:
    cleaned = url.strip()
    parsed = urlparse(cleaned)

    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        raise ExtractionError("http または https のURLを入力してください。")

    if os.getenv("ALLOW_PRIVATE_URLS", "").lower() in {"1", "true", "yes"}:
        return cleaned

    hostname = parsed.hostname
    if not hostname:
        raise ExtractionError("URLのホスト名を確認できません。")

    try:
        addresses = socket.getaddrinfo(hostname, None)
    except socket.gaierror as exc:
        raise ExtractionError("ホスト名を解決できませんでした。") from exc

    for address in addresses:
        ip = ipaddress.ip_address(address[4][0])
        if ip.is_private or ip.is_loopback or ip.is_link_local or ip.is_reserved:
            raise ExtractionError("公開WebページのURLを入力してください。")

    return cleaned


async def fetch_html(url: str) -> tuple[str, str]:
    safe_url = validate_fetch_url(url)
    headers = {
        "User-Agent": (
            "Mozilla/5.0 (compatible; VideoURLExtractor/0.1; "
            "+https://example.dev/video-url-extractor)"
        )
    }

    async with httpx.AsyncClient(
        timeout=REQUEST_TIMEOUT,
        follow_redirects=True,
        headers=headers,
    ) as client:
        async with client.stream("GET", safe_url) as response:
            response.raise_for_status()

            content_type = response.headers.get("content-type", "")
            if "text/html" not in content_type and "application/xhtml+xml" not in content_type:
                raise ExtractionError("HTMLページのURLを入力してください。")

            chunks: list[bytes] = []
            total = 0
            async for chunk in response.aiter_bytes():
                total += len(chunk)
                if total > MAX_HTML_BYTES:
                    raise ExtractionError("HTMLが大きすぎるため解析を中止しました。")
                chunks.append(chunk)

    return b"".join(chunks).decode(response.encoding or "utf-8", errors="replace"), str(response.url)


def extract_video_urls(html_text: str, page_url: str) -> list[dict[str, str]]:
    soup = BeautifulSoup(html_text, "html.parser")
    candidates: list[VideoCandidate] = []

    candidates.extend(_extract_from_media_tags(soup, page_url))
    candidates.extend(_extract_from_metadata(soup, page_url))
    candidates.extend(_extract_from_embeds(soup, page_url))
    candidates.extend(_extract_from_text(html_text, page_url))

    deduped: dict[str, VideoCandidate] = {}
    for candidate in candidates:
        if candidate.url not in deduped:
            deduped[candidate.url] = candidate

    return [asdict(candidate) for candidate in deduped.values()]


def _extract_from_media_tags(soup: BeautifulSoup, page_url: str) -> Iterable[VideoCandidate]:
    for tag in soup.find_all(list(DIRECT_MEDIA_TAGS)):
        tag_name = tag.name or "media"
        is_video_context = tag_name == "video" or tag.find_parent("video") is not None
        declared_type = str(tag.get("type", "")).lower()

        for attr in MEDIA_ATTRS:
            value = tag.get(attr)
            if not value:
                continue

            normalized = normalize_url(str(value), page_url)
            if normalized and (
                _is_video_file(normalized)
                or is_video_context
                or declared_type.startswith("video/")
                or "mpegurl" in declared_type
            ):
                yield VideoCandidate(
                    url=normalized,
                    kind=_kind_for_url(normalized),
                    source=f"<{tag_name}> {attr}",
                )


def _extract_from_metadata(soup: BeautifulSoup, page_url: str) -> Iterable[VideoCandidate]:
    selectors = [
        ("meta", {"property": re.compile(r"^(og:video|og:video:url|og:video:secure_url)$", re.I)}),
        ("meta", {"name": re.compile(r"^(twitter:player:stream|twitter:player)$", re.I)}),
        ("link", {"rel": re.compile(r"(preload|canonical)", re.I)}),
    ]

    for name, attrs in selectors:
        for tag in soup.find_all(name, attrs=attrs):
            for attr in MEDIA_ATTRS:
                value = tag.get(attr)
                if not value:
                    continue

                normalized = normalize_url(str(value), page_url)
                if normalized and _is_video_file(normalized):
                    yield VideoCandidate(
                        url=normalized,
                        kind=_kind_for_url(normalized),
                        source=f"<{name}> {attr}",
                    )


def _extract_from_embeds(soup: BeautifulSoup, page_url: str) -> Iterable[VideoCandidate]:
    for tag in soup.find_all(list(EMBED_TAGS)):
        tag_name = tag.name or "embed"
        for attr in ("src", "data", "data-src", "data-url"):
            value = tag.get(attr)
            normalized = normalize_url(str(value), page_url) if value else None
            if not normalized:
                continue

            yield VideoCandidate(
                url=normalized,
                kind="embed" if not _is_video_file(normalized) else _kind_for_url(normalized),
                source=f"<{tag_name}> {attr}",
            )


def _extract_from_text(html_text: str, page_url: str) -> Iterable[VideoCandidate]:
    searchable = html.unescape(html_text).replace("\\/", "/")

    for regex in (ABSOLUTE_VIDEO_RE, RELATIVE_VIDEO_RE):
        for match in regex.finditer(searchable):
            normalized = normalize_url(match.group("url"), page_url)
            if normalized and _is_video_file(normalized):
                yield VideoCandidate(
                    url=normalized,
                    kind=_kind_for_url(normalized),
                    source="HTML text scan",
                )


def normalize_url(raw_value: str, page_url: str) -> str | None:
    value = html.unescape(raw_value).strip()
    value = value.replace("\\/", "/").replace("\\u002F", "/").replace("\\u002f", "/")
    value = value.strip(" \t\r\n\"'`")

    if value.startswith("url(") and value.endswith(")"):
        value = value[4:-1].strip(" \t\r\n\"'")

    value = _trim_trailing_noise(value)

    if not value:
        return None

    parsed = urlparse(value)
    if parsed.scheme in {"data", "blob", "javascript", "mailto", "tel"}:
        return None

    if value.startswith("//"):
        return f"{urlparse(page_url).scheme}:{value}"

    return urljoin(page_url, value)


def _trim_trailing_noise(value: str) -> str:
    return value.rstrip(".,;:)]}\"'")


def _is_video_file(url: str) -> bool:
    path = urlparse(url).path.lower()
    return any(path.endswith(f".{extension}") for extension in VIDEO_EXTENSIONS)


def _kind_for_url(url: str) -> str:
    path = urlparse(url).path.lower()
    if path.endswith(".m3u8"):
        return "hls"
    return "file"

File: src/extractor.js
const VIDEO_EXTENSIONS = ["mp4", "m3u8", "webm", "mov", "m4v", "ogv"];
const IMAGE_EXTENSIONS = ["jpg", "jpeg", "png", "webp", "avif", "gif"];
const NON_PAGE_EXTENSIONS = [
  "7z",
  "avi",
  "css",
  "csv",
  "doc",
  "docx",
  "exe",
  "gz",
  "ico",
  "jpeg",
  "jpg",
  "js",
  "json",
  "m4v",
  "mov",
  "mp3",
  "mp4",
  "pdf",
  "png",
  "rar",
  "svg",
  "ts",
  "txt",
  "webm",
  "webp",
  "xls",
  "xlsx",
  "zip",
];
const THUMBNAIL_KEYS = [
  "cover",
  "coverimage",
  "image",
  "poster",
  "previewurl",
  "thumbnail",
  "thumbnailurl",
];
const VILOLO_PAGE_HOST_RE =
  /^(video\.twimg-image\.com|video\.twimgx\.com|cdn\d+\.(mvfile\.com|twimg-media\.com|image-share\.cc))$/;
const VILOLO_SHORT_LINK_RE = /^[A-Za-z0-9_-]{4,64}$/;
const VILOLO_GENERIC_SHORT_LINK_RE = /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z0-9_-]{4,64}$/;
const VILOLO_API_BASE = "https://rwzugqnp.fun800.click/app-api";
const RELATED_VIDEO_PAGE_SIZE = 20;
const RELATED_VIDEO_MAX_PAGES = 5;
const MAX_HTML_CHARS = 5 * 1024 * 1024;
const MEDIA_ATTRS = [
  "src",
  "href",
  "data-src",
  "data-url",
  "data-video",
  "data-video-url",
  "data-hls",
  "data-file",
  "content",
];

const TAG_RE = /<(?<tag>video|source|iframe|embed|object|meta|link|a)\b(?<attrs>[^>]*)>/gi;
const ANCHOR_RE = /<a\b(?<attrs>[^>]*)>/gi;
const ATTR_RE = /([\w:-]+)\s*=\s*("([^"]*)"|'([^']*)'|([^\s"'=<>`]+))/gi;
const ABSOLUTE_VIDEO_RE = new RegExp(
  `((?:https?:)?//[^\\s"'<>]+?\\.(?:${VIDEO_EXTENSIONS.join("|")})(?:\\?[^\\s"'<>]*)?)`,
  "gi",
);
const RELATIVE_VIDEO_RE = new RegExp(
  `((?:/|\\.\\.?/)[^\\s"'<>]+?\\.(?:${VIDEO_EXTENSIONS.join("|")})(?:\\?[^\\s"'<>]*)?)`,
  "gi",
);

export async function readJsonBody(request) {
  const contentType = request.headers.get("content-type") || "";
  if (!contentType.includes("application/json")) {
    throw httpError("Content-Typeはapplication/jsonにしてください。", 415);
  }

  try {
    return await request.json();
  } catch {
    throw httpError("JSONの形式を確認してください。", 400);
  }
}

export function validatePageUrl(value) {
  const rawUrl = String(value || "").trim();
  let url;

  try {
    url = new URL(rawUrl);
  } catch {
    throw httpError("http または https のURLを入力してください。", 400);
  }

  if (!["http:", "https:"].includes(url.protocol)) {
    throw httpError("http または https のURLを入力してください。", 400);
  }

  const host = url.hostname.toLowerCase();
  if (
    host === "localhost" ||
    host.endsWith(".localhost") ||
    host === "0.0.0.0" ||
    host === "127.0.0.1" ||
    host === "::1" ||
    host === "[::1]" ||
    host.startsWith("10.") ||
    host.startsWith("192.168.") ||
    /^172\.(1[6-9]|2\d|3[0-1])\./.test(host)
  ) {
    throw httpError("公開WebページのURLを入力してください。", 400);
  }

  return url.toString();
}

export function validateDownloadUrl(value) {
  const url = validatePageUrl(value);

  if (!isVideoFile(url)) {
    throw httpError("保存できる動画URLではありません。", 400);
  }

  return url;
}

export function isHlsUrl(url) {
  return new URL(url).pathname.toLowerCase().endsWith(".m3u8");
}

export function getDownloadFileName(url) {
  const parsed = new URL(url);
  const lastSegment = parsed.pathname.split("/").filter(Boolean).pop();
  return sanitizeFileName(safeDecodeURIComponent(lastSegment || "video"));
}

export function getTransportStreamFileName(url) {
  return getDownloadFileName(url).replace(/\.[^.]+$/, "") + ".ts";
}

export function parseHlsPlaylist(text, playlistUrl) {
  const lines = String(text || "")
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);

  if (lines.some((line) => line.startsWith("#EXT-X-KEY"))) {
    throw httpError("暗号化されたHLSは保存できません。", 400);
  }

  if (lines.some((line) => line.startsWith("#EXT-X-MAP"))) {
    throw httpError("このHLS形式は保存に対応していません。", 400);
  }

  const variants = [];
  const segments = [];

  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index];
    if (line.startsWith("#EXT-X-STREAM-INF")) {
      const nextLine = lines[index + 1];
      if (nextLine && !nextLine.startsWith("#")) {
        variants.push({
          url: new URL(nextLine, playlistUrl).toString(),
          bandwidth: Number(line.match(/BANDWIDTH=(\d+)/)?.[1] || 0),
        });
      }
      continue;
    }

    if (!line.startsWith("#")) {
      segments.push(new URL(line, playlistUrl).toString());
    }
  }

  variants.sort((a, b) => b.bandwidth - a.bandwidth);

  return {
    variants,
    segments: variants.length ? [] : segments,
  };
}

export async function fetchHtml(url) {
  const response = await fetch(url, {
    headers: {
      "User-Agent":
        "Mozilla/5.0 (compatible; VideoURLExtractor/1.0; +https://pages.dev)",
      Accept: "text/html,application/xhtml+xml",
    },
    redirect: "follow",
  });

  if (!response.ok) {
    throw httpError(`取得先ページがHTTP ${response.status} を返しました。`, 502);
  }

  const contentType = response.headers.get("content-type") || "";
  if (
    !contentType.includes("text/html") &&
    !contentType.includes("application/xhtml+xml")
  ) {
    throw httpError("HTMLページのURLを入力してください。", 400);
  }

  const html = await response.text();
  if (html.length > MAX_HTML_CHARS) {
    throw httpError("HTMLが大きすぎるため解析を中止しました。", 400);
  }

  return {
    html,
    pageUrl: response.url || url,
  };
}

export function extractVideoUrls(html, pageUrl) {
  const candidates = [
    ...extractFromTags(html, pageUrl),
    ...extractFromText(html, pageUrl),
  ];
  const deduped = new Map();

  for (const candidate of candidates) {
    if (!deduped.has(candidate.url)) {
      deduped.set(candidate.url, candidate);
    }
  }

  return [...deduped.values()];
}

export async function extractKnownProviderUrls(pageUrl, fetchImpl = fetch) {
  const url = new URL(pageUrl);
  const shortLink = getViloloLikeShortLink(url);
  if (!shortLink) {
    return [];
  }

  const payload = await fetchViloloJson(
    "/flow/land-page/getInfo",
    {
      externalLinks: shortLink,
      domain: url.hostname,
    },
    fetchImpl,
  );

  if (!payload) {
    return [];
  }

  const currentResults = extractFromJson(payload, pageUrl, "Vilolo media API");
  const relatedResults = await extractViloloRelatedUrls(
    payload,
    url.hostname,
    pageUrl,
    fetchImpl,
  );

  return dedupeCandidates([...currentResults, ...relatedResults]);
}

function getViloloLikeShortLink(url) {
  const pathSegments = url.pathname.split("/").filter(Boolean);
  if (pathSegments.length !== 1 || url.search || url.hash) {
    return "";
  }

  const shortLink = pathSegments[0];
  if (!VILOLO_SHORT_LINK_RE.test(shortLink)) {
    return "";
  }

  if (VILOLO_PAGE_HOST_RE.test(url.hostname)) {
    return shortLink;
  }

  return VILOLO_GENERIC_SHORT_LINK_RE.test(shortLink) &&
    looksLikeShortShareHost(url.hostname)
    ? shortLink
    : "";
}

function looksLikeShortShareHost(hostname) {
  if (!hostname.includes(".")) {
    return false;
  }

  const labels = hostname.toLowerCase().split(".");
  if (labels.some((label) => ["localhost", "local"].includes(label))) {
    return false;
  }

  // Vilolo-like mirrors use compact one-path IDs on ordinary public hosts.
  // The provider API validates the domain/link pair; a miss simply returns no data.
  return labels.length >= 2 && labels.every((label) => /^[a-z0-9-]{1,63}$/.test(label));
}

async function extractViloloRelatedUrls(payload, domain, pageUrl, fetchImpl) {
  const externalLinks = getNestedValue(payload, [
    "data",
    "info",
    "extraInfo",
    "externalLinks",
  ]);
  if (!externalLinks) {
    return [];
  }

  const sortOrder =
    getNestedValue(payload, ["data", "info", "extraInfo", "sortOrder"]) || "3";
  const relatedPayloads = [];

  for (let pageNo = 1; pageNo <= RELATED_VIDEO_MAX_PAGES; pageNo += 1) {
    const relatedPayload = await fetchViloloJson(
      "/flow/land-page/list_by_links_page",
      {
        externalLinks,
        domain,
        pageNo: String(pageNo),
        pageSize: String(RELATED_VIDEO_PAGE_SIZE),
        sortOrder,
      },
      fetchImpl,
    );

    const items = Array.isArray(relatedPayload?.data?.list)
      ? relatedPayload.data.list
      : [];
    if (!items.length) {
      break;
    }

    relatedPayloads.push(relatedPayload);

    const total = Number(relatedPayload?.data?.total || 0);
    if (total > 0 && pageNo * RELATED_VIDEO_PAGE_SIZE >= total) {
      break;
    }
  }

  return extractFromJson(relatedPayloads, pageUrl, "Vilolo related API")
    .map((result) => ({
      ...result,
      sourcePage: buildViloloLandingPageUrl(domain, result.landingPage) || pageUrl,
    }))
    .map(({ landingPage, ...result }) => result);
}

async function fetchViloloJson(path, params, fetchImpl) {
  const apiUrl = new URL(`${VILOLO_API_BASE}${path}`);
  Object.entries(params).forEach(([key, value]) => {
    apiUrl.searchParams.set(key, value);
  });

  const response = await fetchImpl(apiUrl, {
    headers: { Accept: "application/json" },
    redirect: "follow",
  });

  if (!response.ok) {
    return null;
  }

  try {
    const payload = await response.json();
    return payload?.code === 0 ? payload : null;
  } catch {
    return null;
  }
}

function getNestedValue(value, path) {
  return path.reduce((current, key) => current?.[key], value) || "";
}

function buildViloloLandingPageUrl(domain, landingPage) {
  if (!landingPage) {
    return "";
  }

  return `https://${domain}/${encodeURIComponent(landingPage)}`;
}

export function extractLinkedPageUrls(html, pageUrl, limit = 8) {
  const links = [];
  const seen = new Set([new URL(pageUrl).toString()]);

  for (const match of html.matchAll(ANCHOR_RE)) {
    const attrs = parseAttrs(match.groups.attrs || "");
    const linkedUrl = normalizeUrl(attrs.href, pageUrl);
    if (!linkedUrl || seen.has(linkedUrl) || shouldSkipLinkedPage(linkedUrl)) {
      continue;
    }

    try {
      validatePageUrl(linkedUrl);
    } catch {
      continue;
    }

    seen.add(linkedUrl);
    links.push(linkedUrl);
    if (links.length >= limit) {
      break;
    }
  }

  return links;
}

function dedupeCandidates(candidates) {
  const deduped = new Map();

  for (const candidate of candidates) {
    if (!deduped.has(candidate.url)) {
      deduped.set(candidate.url, candidate);
      continue;
    }

    const existing = deduped.get(candidate.url);
    deduped.set(candidate.url, {
      ...existing,
      ...Object.fromEntries(
        Object.entries(candidate).filter(([, value]) => value && value !== ""),
      ),
    });
  }

  return [...deduped.values()];
}

function extractFromTags(html, pageUrl) {
  const candidates = [];

  for (const match of html.matchAll(TAG_RE)) {
    const tag = match.groups.tag.toLowerCase();
    const attrs = parseAttrs(match.groups.attrs || "");
    const type = (attrs.type || "").toLowerCase();
    const thumbnailUrl =
      tag === "video" ? normalizeThumbnailUrl(attrs.poster, pageUrl) : null;
    const isDirectMediaTag =
      tag === "video" ||
      (tag === "source" &&
        (type.startsWith("video/") || type.includes("mpegurl") || type === ""));
    const isEmbedTag = ["iframe", "embed", "object"].includes(tag);

    for (const attr of MEDIA_ATTRS) {
      const value = attrs[attr];
      if (!value) continue;

      const normalized = normalizeUrl(value, pageUrl);
      if (!normalized) continue;

      if (
        isVideoFile(normalized) ||
        isDirectMediaTag ||
        (isEmbedTag && ["src", "data", "data-src", "data-url"].includes(attr))
      ) {
        candidates.push({
          url: normalized,
          kind: isVideoFile(normalized) ? kindForUrl(normalized) : "embed",
          source: `<${tag}> ${attr}`,
          ...(thumbnailUrl ? { thumbnailUrl } : {}),
        });
      }
    }
  }

  return candidates;
}

function extractFromJson(value, pageUrl, source) {
  const candidates = [];
  const seen = new Set();
  visitJson(value, "", "");
  return candidates;

  function visitJson(item, inheritedThumbnailUrl, inheritedLandingPage) {
    if (typeof item === "string") {
      const normalized = normalizeUrl(item, pageUrl);
      if (normalized && isVideoFile(normalized) && !seen.has(normalized)) {
        seen.add(normalized);
        candidates.push({
          url: normalized,
          kind: kindForUrl(normalized),
          source,
          ...(inheritedThumbnailUrl
            ? { thumbnailUrl: inheritedThumbnailUrl }
            : {}),
          ...(inheritedLandingPage ? { landingPage: inheritedLandingPage } : {}),
        });
      }
      return;
    }

    if (Array.isArray(item)) {
      item.forEach((child) =>
        visitJson(child, inheritedThumbnailUrl, inheritedLandingPage),
      );
      return;
    }

    if (item && typeof item === "object") {
      const thumbnailUrl = findThumbnailUrl(item, pageUrl) || inheritedThumbnailUrl;
      const landingPage =
        typeof item.landingPage === "string"
          ? item.landingPage
          : inheritedLandingPage;
      Object.values(item).forEach((child) =>
        visitJson(child, thumbnailUrl, landingPage),
      );
    }
  }
}

function findThumbnailUrl(item, pageUrl) {
  for (const [key, value] of Object.entries(item)) {
    if (!THUMBNAIL_KEYS.includes(key.toLowerCase()) || typeof value !== "string") {
      continue;
    }

    const normalized = normalizeThumbnailUrl(value, pageUrl);
    if (normalized) {
      return normalized;
    }
  }

  return "";
}

function extractFromText(html, pageUrl) {
  const candidates = [];
  const searchable = decodeHtmlEntities(html)
    .replaceAll("\\/", "/")
    .replaceAll("\\u002F", "/")
    .replaceAll("\\u002f", "/");

  for (const regex of [ABSOLUTE_VIDEO_RE, RELATIVE_VIDEO_RE]) {
    for (const match of searchable.matchAll(regex)) {
      const normalized = normalizeUrl(match[1], pageUrl);
      if (!normalized || !isVideoFile(normalized)) continue;

      candidates.push({
        url: normalized,
        kind: kindForUrl(normalized),
        source: "HTML text scan",
      });
    }
  }

  return candidates;
}

function parseAttrs(rawAttrs) {
  const attrs = {};

  for (const match of rawAttrs.matchAll(ATTR_RE)) {
    attrs[match[1].toLowerCase()] = decodeHtmlEntities(
      match[3] ?? match[4] ?? match[5] ?? "",
    );
  }

  return attrs;
}

export function normalizeUrl(rawValue, pageUrl) {
  let value = decodeHtmlEntities(String(rawValue || ""))
    .trim()
    .replaceAll("\\/", "/")
    .replaceAll("\\u002F", "/")
    .replaceAll("\\u002f", "/")
    .replace(/^url\((.*)\)$/i, "$1")
    .trim()
    .replace(/^[ "'`]+|[ "'`]+$/g, "")
    .replace(/[.,;:)\]}"']+$/g, "");

  if (!value) return null;
  if (/^(data|blob|javascript|mailto|tel):/i.test(value)) return null;

  try {
    if (value.startsWith("//")) {
      return `${new URL(pageUrl).protocol}${value}`;
    }

    return new URL(value, pageUrl).toString();
  } catch {
    return null;
  }
}

function decodeHtmlEntities(value) {
  return value
    .replace(/&amp;/g, "&")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">");
}

function isVideoFile(url) {
  const path = new URL(url).pathname.toLowerCase();
  return VIDEO_EXTENSIONS.some((extension) => path.endsWith(`.${extension}`));
}

function normalizeThumbnailUrl(rawValue, pageUrl) {
  const normalized = normalizeUrl(rawValue, pageUrl);
  if (!normalized) return "";

  const path = new URL(normalized).pathname.toLowerCase();
  return IMAGE_EXTENSIONS.some((extension) => path.endsWith(`.${extension}`))
    ? normalized
    : "";
}

function shouldSkipLinkedPage(url) {
  const parsed = new URL(url);
  if (!["http:", "https:"].includes(parsed.protocol)) return true;

  const extension = parsed.pathname.split(".").pop()?.toLowerCase() || "";
  return NON_PAGE_EXTENSIONS.includes(extension);
}

function kindForUrl(url) {
  return isHlsUrl(url) ? "hls" : "file";
}

function sanitizeFileName(value) {
  const cleaned = String(value || "video").replace(/[\\/:*?"<>|]/g, "_");
  return cleaned || "video";
}

function safeDecodeURIComponent(value) {
  try {
    return decodeURIComponent(String(value || ""));
  } catch {
    return String(value || "");
  }
}

export function jsonResponse(payload, status = 200) {
  return new Response(JSON.stringify(payload), {
    status,
    headers: {
      "Content-Type": "application/json; charset=utf-8",
      "Cache-Control": "no-store",
    },
  });
}

function httpError(message, status) {
  const error = new Error(message);
  error.status = status;
  return error;
}
